src/std/foreign.ss

```diff
@@ -43,10 +44,172 @@
+                                       (string-append struct-str "_ffi_free")
+                                       "ffi_free")))
+               (string-compat-types (if string-compat-required?
+                                      `((c-declare ,default-free-body)
+                                        (c-define-type ,shallow-ptr
+                                              (pointer ,struct (,struct-ptr) "ffi_free"))))))
+          `(begin
+             ,@(append `((c-define-type ,struct (struct ,struct-str))
+                         (c-define-type ,struct-ptr
+                                        (pointer ,struct (,struct-ptr) ,release-function))
+                         (c-define-type ,borrowed-ptr (pointer ,struct (,struct-ptr))))
+                       string-compat-types)
+
+
+                  (define ,(string->symbol (string-append struct-str "-ptr?"))
+                    (lambda (obj)
+                      (and (foreign? obj)
+                         (equal? (foreign-tags obj) (quote (,struct-ptr))))))
+
+                  ;; getter and setters
+                  ,@(apply append
+                     (map (lambda (m)
+                            (let* ((member-name (symbol->string (car m)))
+                                   (member-type (cdr m))
+                                   (getter-name (string-append struct-str "-" member-name))
+                                   (setter-body (cond
+                                                 ((member member-type string-types)
+                                                  (string-setter-body member-name))
+                                                 (else
+                                                  (string-append
+                                                   "___arg1->" member-name " = ___arg2;" "\n"
+                                                   "___return;" "\n")))))
+                              `((define ,(string->symbol getter-name)
+                                  (c-lambda (,struct-ptr) ,member-type
+                                       ,(string-append
+                                         "___return(___arg1->" member-name ");")))
+
+                                (define ,(string->symbol (string-append getter-name "-set!"))
+                                  (c-lambda (,struct-ptr ,member-type) void
+                                            ,setter-body)))))
+                          members))
+
+                  ;; malloc
+                  (define ,(string->symbol (string-append "malloc-" struct-str))
+                    (c-lambda () ,struct-ptr
+                         ,(string-append
+                           "struct " struct-str "* var = malloc(sizeof(struct " struct-str "));" "\n"
+                           "memset(var, 0, sizeof(struct " struct-str "));"
+                          "if (var == NULL)" "\n"
+                          "    ___return (NULL);" "\n"
```

Comment:
need to check for `NULL` _before_ memsetting, otherwise it will segfault if the allocation fails.